File: java/SearchinRotatedSortedArray.java
/*
 * Filename: SearchinRotatedSortedArray.java
 * Author: Jun Xu
 * Email: junx@bu.edu
 * Created Time: Sun 22 Mar 2015 01:24:12 PM EDT
 */

public class Solution {
	public int search(int[] A, int target) {
		int n = A.length;

		if (n == 0) {
			return -1;
		}

		if (n == 1) {
			if (A[0] == target) {
				return 0;
			} else {
				return -1;
			}
		}

		int i = 1;
		while (i < n && A[i-1] <= A[i]) {
			i++;
		}

		int head = 0;
		int tail = n - 1;
		int mid;
		int val;
		
		while (head <= tail) {
			mid = (head + tail) / 2;
			val = A[(mid + i) % n];
			if (val == target) {
				return mid;
			} else if (val < target) {
				head = mid + 1;
			} else {
				tail = mid - 1;
			}
		}

		return -1;
	}
}

File: cpp/search_in_rotated_sorted_array.cc
/*
 * Filename: search_in_rotated_sorted_array.cc
 * Author: Jun Xu
 * Email: jxu.vimacs@gmail.com
 * Created Time: Sat 25 Jul 2015 12:29:03 PM EDT
 */

#include <iostream>
#include <vector>

using namespace std;

class Solution {
	public:
		int search(vector<int>& nums, int target) {
			if (nums.size() == 1) {
				return nums[0] == target ? 0 : -1;
			}

			int len = nums.size();
			int offset = 0;
			int stt = 0;
			int end = nums.size() - 1;
			int mid = -1;

			for (int i = 1; i < len; i++) {
				if (nums[i] < nums[i-1]) {
					offset = i;
				}
			}

			while (stt <= end) {
				mid = (stt + end) / 2;
				int idx = (mid + offset) % len;
				
				if (nums[idx] == target) {
					return idx;
				} else if (nums[idx] > target) {
					end = mid - 1;
				} else {
					stt = mid + 1;
				}
			}

			return -1;
		}
};
